File: interviews/A/VO/LRU.py
class LinkedNode:
    
    def __init__(self, key=None, val=None, next=None):
        self.key = key 
        self.val = val 
        self.next = next 

class LRUCache:
    """
    @param: capacity: An integer
    """
    def __init__(self, capacity):
        
        self.capacity = capacity
        
        self.head = LinkedNode()
        
        self.tail = self.head
        
        self.key_to_prev = {}
        
    def popFront(self):
        
        del self.key_to_prev[self.head.next.key]
        
        self.head.next = self.head.next.next 
        
        self.key_to_prev[self.head.next.key] = self.head 
        
    def pushBack(self, node):
        
        self.key_to_prev[node.key] = self.tail 
        
        self.tail.next = node 
        
        self.tail = self.tail.next 
        
    def kick(self, prev):
        
        node = prev.next 
        
        if node == self.tail:
            
            return 
        
        prev.next = node.next 
        
        if prev.next is not None:
            
            self.key_to_prev[prev.next.key] = prev 
            
            node.next = None 
            
        self.pushBack(node)
            
        
        
        

    """
    @param: key: An integer
    @return: An integer
    """
    def get(self, key):
        # write your code here
        
        if key not in self.key_to_prev:
            
            return -1 
            
        prev = self.key_to_prev[key] 
        
        node = prev.next 
        
        self.kick(prev)
        
        return node.val
        
        

    """
    @param: key: An integer
    @param: value: An integer
    @return: nothing
    """
    def set(self, key, value):
        # write your code here
        
        if key in self.key_to_prev:
            
            prev = self.key_to_prev[key]
            
            self.kick(prev)
            
            self.key_to_prev[key].next.val = value 
            
        else:
            
            node = LinkedNode(key, value)
            
            self.pushBack(node)
            
            if len(self.key_to_prev) > self.capacity:
                
                self.popFront()
        
        
        
            
        
        
        
        
                

File: interviews/MobiTV/LRU.java
class LinkedNode {
    
    int key; 
    int value; 
    LinkedNode next; 
    
    public LinkedNode(int key, int value) {
        
        this.key = key;
        this.value = value; 
        this.next = null;
    }
}

public class LRUCache {
    
    Map<Integer, LinkedNode> keyToPrev; 
    LinkedNode head;
    LinkedNode tail;
    int capacity;
    
    /*
    * @param capacity: An integer
    */public LRUCache(int capacity) {
        // do intialization if necessary
        this.keyToPrev = new HashMap<Integer, LinkedNode>();
        this.head = new LinkedNode(-1, -1);
        this.tail = this.head; 
        this.capacity = capacity;
        
    }
    
    private void popFront() {
        
        this.keyToPrev.remove(this.head.next.key);
        this.head.next = this.head.next.next;
        this.keyToPrev.put(this.head.next.key, this.head); 
    }
    
    private void pushBack(LinkedNode node) {
        
         
        this.keyToPrev.put(node.key, this.tail); 
        this.tail.next = node;
        this.tail = this.tail.next;
    }
    
    private void kick(LinkedNode prev) {
        
        LinkedNode node = prev.next;
        if (node == this.tail) 
            return;
            
        prev.next = node.next; 
        
        if (node.next != null) {
            
            this.keyToPrev.put(node.next.key, prev);
            node.next = null;
        }
        
        this.pushBack(node);
    }
    
    
    /*
     * @param key: An integer
     * @return: An integer
     */
    public int get(int key) {
        // write your code here
        
        if (!this.keyToPrev.containsKey(key)) {
            return -1; 
        }
        
        LinkedNode prev = this.keyToPrev.get(key); 
        LinkedNode node = prev.next;
        
        this.kick(prev);
        return node.value;
    }

    /*
     * @param key: An integer
     * @param value: An integer
     * @return: nothing
     */
    public void set(int key, int value) {
        // write your code here
        
        if (this.keyToPrev.containsKey(key)) {
            
            LinkedNode prev = this.keyToPrev.get(key);
            
            this.kick(prev);
            this.keyToPrev.get(key).next.value = value;
            
        }
        
        else {
            
            LinkedNode node = new LinkedNode(key, value); 
            
            this.pushBack(node);
            
            if (this.keyToPrev.size() > this.capacity) {
                this.popFront();
            }
        }
        
    }
}